--- a/drivers/input/input.c
+++ b/drivers/input/input.c
@@ -354,11 +354,5 @@
 	}
 }
 
-#ifdef CONFIG_KSU_SUSFS
-extern bool ksu_input_hook __read_mostly;
-extern __attribute__((cold)) int ksu_handle_input_handle_event(
-			unsigned int *type, unsigned int *code, int *value);
-#endif
-
 void input_handle_event(struct input_dev *dev,
 			unsigned int type, unsigned int code, int value)
@@ -368,10 +362,6 @@
 	lockdep_assert_held(&dev->event_lock);
 
 	disposition = input_get_disposition(dev, type, code, &value);
-#ifdef CONFIG_KSU_SUSFS
-	if (unlikely(ksu_input_hook))
-		ksu_handle_input_handle_event(&type, &code, &value);
-#endif
 	if (disposition != INPUT_IGNORE_EVENT) {
 		if (type != EV_SYN)
 			add_input_randomness(type, code, value);
--- a/fs/devpts/inode.c
+++ b/fs/devpts/inode.c
@@ -580,10 +580,6 @@
 	return dentry;
 }
 
-#ifdef CONFIG_KSU_SUSFS
-extern int ksu_handle_devpts(struct inode*);
-#endif
-
 /**
  * devpts_get_priv -- get private data for a slave
  * @dentry: dentry of the slave
@@ -592,14 +588,6 @@
  */
 void *devpts_get_priv(struct dentry *dentry)
 {
-#ifdef CONFIG_KSU_SUSFS
-	if (likely(susfs_is_current_proc_umounted())) {
-		goto orig_flow;
-	}
-	ksu_handle_devpts(dentry->d_inode);
-orig_flow:
-#endif
-
 	if (dentry->d_sb->s_magic != DEVPTS_SUPER_MAGIC)
 		return NULL;
 	return dentry->d_fsdata;
--- a/fs/exec.c
+++ b/fs/exec.c
@@ -1913,16 +1913,5 @@
 	return retval;
 }
 
-#ifdef CONFIG_KSU_SUSFS
-extern bool ksu_execveat_hook __read_mostly;
-extern bool ksu_su_compat_enabled __read_mostly;
-extern bool susfs_is_sdcard_android_data_decrypted __read_mostly;
-extern bool __ksu_is_allow_uid_for_current(uid_t uid);
-extern int ksu_handle_execveat(int *fd, struct filename **filename_ptr, void *argv,
-			void *envp, int *flags);
-extern int ksu_handle_execveat_sucompat(int *fd, struct filename **filename_ptr, void *argv,
-				void *envp, int *flags);
-#endif
-
 static int do_execveat_common(int fd, struct filename *filename,
 			      struct user_arg_ptr argv,
@@ -1935,20 +1924,6 @@
 	if (IS_ERR(filename))
 		return PTR_ERR(filename);
 
-#ifdef CONFIG_KSU_SUSFS
-	if (likely(susfs_is_current_proc_umounted()) || !ksu_su_compat_enabled) {
-		goto orig_flow;
-	}
-
-	if (unlikely(ksu_execveat_hook || !susfs_is_sdcard_android_data_decrypted)) {
-		ksu_handle_execveat(&fd, &filename, &argv, &envp, &flags);
-	} else if ((__ksu_is_allow_uid_for_current(current_uid().val))) {
-		ksu_handle_execveat_sucompat(&fd, &filename, &argv, &envp, &flags);
-	}
-
-orig_flow:
-#endif
-
 	/*
 	 * We move the actual failure in case of RLIMIT_NPROC excess from
 	 * set*uid() to execve() because too many poorly written programs
--- a/fs/read_write.c
+++ b/fs/read_write.c
@@ -718,18 +718,8 @@
 	return ret;
 }
 
-#ifdef CONFIG_KSU_SUSFS
-extern bool ksu_init_rc_hook __read_mostly;
-extern __attribute__((cold)) void ksu_handle_sys_read(unsigned int fd);
-#endif
-
 SYSCALL_DEFINE3(read, unsigned int, fd, char __user *, buf, size_t, count)
 {
-#ifdef CONFIG_KSU_SUSFS
-	if (unlikely(ksu_init_rc_hook))
-		ksu_handle_sys_read(fd);
-#endif
-
 	return ksys_read(fd, buf, count);
 }
 
--- a/fs/stat.c
+++ b/fs/stat.c
@@ -242,6 +242,6 @@
 
 #ifdef CONFIG_KSU_SUSFS
-extern bool ksu_init_rc_hook __read_mostly;
+extern bool ksu_vfs_read_hook __read_mostly;
 extern void ksu_handle_vfs_fstat(int fd, loff_t *kstat_size_ptr);
 #endif // #ifdef CONFIG_KSU_SUSFS
 
@@ -257,6 +257,6 @@
 	error = vfs_getattr(&fd_file(f)->f_path, stat, STATX_BASIC_STATS, 0);
 #ifdef CONFIG_KSU_SUSFS
-	if (unlikely(ksu_init_rc_hook)) {
+	if (unlikely(ksu_vfs_read_hook)) {
 		ksu_handle_vfs_fstat(fd, &stat->size);
 	}
 #endif // #ifdef CONFIG_KSU_SUSFS
--- a/kernel/sys.c
+++ b/kernel/sys.c
@@ -678,9 +678,6 @@
 	return __sys_setuid(uid);
 }
 
-#ifdef CONFIG_KSU_SUSFS
-extern int ksu_handle_setresuid(uid_t ruid, uid_t euid, uid_t suid);
-#endif
 
 /*
  * This function implements a generic ability to update ruid, euid,
@@ -695,10 +692,6 @@
 	kuid_t kruid, keuid, ksuid;
 	bool ruid_new, euid_new, suid_new;
 
-#ifdef CONFIG_KSU_SUSFS
-	(void)ksu_handle_setresuid(ruid, euid, suid);
-#endif
-
 	kruid = make_kuid(ns, ruid);
 	keuid = make_kuid(ns, euid);
 	ksuid = make_kuid(ns, suid);
--- a/fs/susfs.c
+++ b/fs/susfs.c
@@ -661,6 +661,8 @@
 	}
 }
 
+bool ksu_su_compat_enabled __attribute__((weak)) = true;
+
 /* susfs_init */
 void susfs_init(void) {
 	SUSFS_LOGI("susfs is initialized! version: " SUSFS_VERSION " \n");
