diff --git a/scripts/mkcompile_h b/scripts/mkcompile_h
index 2596f78e5..dcc5e4bca 100755
--- a/scripts/mkcompile_h
+++ b/scripts/mkcompile_h
@@ -21,7 +21,7 @@ LD_VERSION=$(LC_ALL=C $LD -v | head -n1 |
 
 cat <<EOF
 #define UTS_MACHINE		"${UTS_MACHINE}"
-#define LINUX_COMPILE_BY	"${LINUX_COMPILE_BY}"
-#define LINUX_COMPILE_HOST	"${LINUX_COMPILE_HOST}"
-#define LINUX_COMPILER		"${CC_VERSION}, ${LD_VERSION}"
+#define LINUX_COMPILE_BY	"kleaf"
+#define LINUX_COMPILE_HOST	"build-host"
+#define LINUX_COMPILER		"Android (14043575, +pgo, +bolt, +lto, +mlgo, based on r536225) clang version 19.0.1 (https://android.googlesource.com/toolchain/llvm-project b3a530ec6537146650e42be89f1089e9a3588460), LLD 19.0.1"
 EOF
