--- a/kernel/supercall/supercall.c
+++ b/kernel/supercall/supercall.c
@@ -107,7 +107,7 @@ int ksu_supercall_reboot_handler(void __user **arg)
 }
 
 // downstream: make sure to pass arg as reference, this can allow us to extend things.
-static int ksu_handle_sys_reboot(int magic1, int magic2, unsigned int cmd, void __user **arg)
+int ksu_handle_sys_reboot(int magic1, int magic2, unsigned int cmd, void __user **arg)
 {
 
     if (magic1 != KSU_INSTALL_MAGIC1)
